File: ios-native/ClubFuoco/Models/ClubEvent.swift
import Foundation

/// An upcoming event at one venue, from public.events (the agentbox RA feed —
/// see EVENTS_INGEST_BRIEF.md). Native mirror of ClubEvent in src/lib/events.ts.
///
/// Distinct from `ExternalEvent`, which models the older, thinner `ra_events`
/// cache the explore feed still reads. Events here are linked to a venue by
/// `club_id` resolved at ingest, so no fuzzy venue-name matching is involved.
/// One billed artist on an event, from `events.lineup`. `id` is RA's artist id
/// — the same key `djs.ra_artist_id` uses — so a credit joins to a DJ page
/// exactly. nil on rows scraped before the lineup field existed.
struct LineupCredit: Decodable, Sendable, Hashable, Identifiable {
    let id: String?
    let name: String

    /// Stable identity for ForEach: two DJs can share neither id nor position,
    /// but an id-less legacy credit still needs a key.
    var key: String { id ?? "name:\(name)" }
}

struct ClubEvent: Decodable, Sendable, Identifiable, Hashable {
    let raEventId: String
    let title: String
    let date: String              // yyyy-MM-dd, the listing day
    let startTime: String?        // ISO instant; already the correct Madrid time
    let venueName: String?
    let promoters: [String]?
    /// Names only, unordered — kept for rows that predate `lineup`.
    let artists: [String]?
    /// Ordered credits with RA artist ids (events.lineup jsonb).
    let lineup: [LineupCredit]?
    let interested: Int?
    let attending: Int?
    let raUrl: String?
    let image: String?            // flyer URL, where the scraper captured one
    let description: String?      // event copy (plain text), where RA had one
    let endTime: String?          // ISO instant; nights routinely cross midnight
    let minimumAge: Int?          // 18/20/21. nil = RA holds NO policy (see below)
    let venueCapacity: String?    // free text, and "0" in half the rows
    /// Free text on the source: "€12–€22", "0", "€", "10€/15€". Modelled now so
    /// the detail sheet can show a real door price, but only through
    /// `entryLabel`, which refuses anything that is not an actual amount.
    let cost: String?

    var id: String { raEventId }

    // No `ticketsURL`. `raUrl` is kept as internal provenance — where the row
    // came from, for support and de-duping — but is never surfaced: the source
    // of the listing is not something the customer is shown.

    /// A door price only when the source actually gave one.
    ///
    /// 60% of rows are "0", "€" or blank, and nothing distinguishes "free" from
    /// "unknown" — so anything without a non-zero digit is treated as unknown
    /// and the chip is simply not shown. Guessing here would misstate what
    /// someone pays at the door.
    var entryLabel: String? {
        guard let raw = cost?.trimmingCharacters(in: .whitespacesAndNewlines),
              !raw.isEmpty else { return nil }
        let digits = raw.compactMap(\.wholeNumberValue)
        guard digits.contains(where: { $0 > 0 }) else { return nil }
        return raw
    }

    /// Capacity only when it is a real number. The column is non-null on 99% of
    /// rows but literally "0" on half of them, which is absence, not a room that
    /// holds nobody.
    var capacityLabel: String? {
        guard let raw = venueCapacity?.trimmingCharacters(in: .whitespacesAndNewlines),
              let n = Int(raw), n > 0 else { return nil }
        return n.formatted(.number.grouping(.automatic))
    }

    /// "23:00 – 06:00", plus whether the end lands on the next day. Median night
    /// here runs six hours and most end between 02:00 and 08:00, so the end time
    /// usually belongs to tomorrow — rendering it bare would read as wrong.
    var timeRange: (label: String, crossesMidnight: Bool)? {
        guard let start = Self.instant(startTime) else { return nil }
        let out = DateFormatter()
        out.timeZone = TimeZone(identifier: "Europe/Madrid")
        out.locale = Locale(identifier: "en_GB")
        out.dateFormat = "HH:mm"
        guard let end = Self.instant(endTime) else {
            return (out.string(from: start), false)
        }
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "Europe/Madrid") ?? .current
        let crosses = !cal.isDate(start, inSameDayAs: end)
        return ("\(out.string(from: start)) – \(out.string(from: end))", crosses)
    }

    static func instant(_ iso: String?) -> Date? {
        guard let iso else { return nil }
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = f.date(from: iso) { return d }
        let plain = ISO8601DateFormatter()
        plain.formatOptions = [.withInternetDateTime]
        return plain.date(from: iso)
    }

    /// ("SUN", "19", "JUL") for the date block. Parsed with a fixed calendar so
    /// the day never shifts with the device timezone.
    var dateParts: (weekday: String, day: String, month: String) {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "Europe/Madrid") ?? .current
        let parser = DateFormatter()
        parser.dateFormat = "yyyy-MM-dd"
        parser.timeZone = cal.timeZone
        parser.locale = Locale(identifier: "en_GB")
        guard let parsed = parser.date(from: date) else { return ("", date, "") }

        let out = DateFormatter()
        out.timeZone = cal.timeZone
        out.locale = Locale(identifier: "en_GB")
        out.dateFormat = "EEE"
        let weekday = out.string(from: parsed).uppercased()
        out.dateFormat = "d"
        let day = out.string(from: parsed)
        out.dateFormat = "MMM"
        return (weekday, day, out.string(from: parsed).uppercased())
    }

    /// "23:00" in Barcelona time — the stored instant is correct, so render it
    /// in the venue's zone rather than the reader's.
    var startLabel: String? {
        guard let startTime else { return nil }
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let parsed = iso.date(from: startTime) ?? {
            let plain = ISO8601DateFormatter()
            plain.formatOptions = [.withInternetDateTime]
            return plain.date(from: startTime)
        }()
        guard let parsed else { return nil }
        let out = DateFormatter()
        out.timeZone = TimeZone(identifier: "Europe/Madrid")
        out.locale = Locale(identifier: "en_GB")
        out.dateFormat = "HH:mm"
        return out.string(from: parsed)
    }

    /// The billed lineup: every DJ on the night, in the order RA lists them,
    /// each with RA's artist id so a credit resolves to a DJ page exactly
    /// rather than by name.
    ///
    /// Falls back to the older `artists` name array (no ids, no order) for rows
    /// scraped before the lineup field existed — those credits still render,
    /// they just match by name.
    var credits: [LineupCredit] {
        if let lineup, !lineup.isEmpty { return lineup }
        return (artists ?? []).map { LineupCredit(id: nil, name: $0) }
    }

    /// Shown on the card; the rest sit behind a "+N".
    var visibleCredits: [LineupCredit] { Array(credits.prefix(6)) }
    var extraCredits: Int { max(0, credits.count - visibleCredits.count) }
}

File: android-native/app/src/main/java/com/clubfuoco/app/models/ClubEvent.kt
package com.clubfuoco.app.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * One billed artist on an event. [id] is RA's artist id — the same key
 * `djs.ra_artist_id` uses — so a credit joins to a DJ page exactly. null on
 * rows scraped before the lineup field existed.
 */
@Serializable
data class LineupCredit(
    val id: String? = null,
    val name: String,
) {
    /**
     * Stable identity for a list key: two DJs can share neither id nor
     * position, but an id-less legacy credit still needs a key.
     */
    val key: String get() = id ?: "name:$name"
}

/**
 * An upcoming event at ONE venue, from `public.events`.
 *
 * Distinct from [ExternalEvent], which models the thinner `ra_events` cache the
 * explore feed reads. Events here are linked by `club_id` resolved at ingest, so
 * no fuzzy venue-name matching is involved.
 */
@Serializable
data class ClubEvent(
    @SerialName("ra_event_id") val raEventId: String,
    val title: String = "",
    /** yyyy-MM-dd, the listing day. */
    val date: String = "",
    /** ISO instant; already the correct Madrid time. */
    @SerialName("start_time") val startTime: String? = null,
    @SerialName("end_time") val endTime: String? = null,
    @SerialName("venue_name") val venueName: String? = null,
    val promoters: List<String>? = null,
    /** Names only, unordered — kept for rows predating `lineup`. */
    val artists: List<String>? = null,
    /** Ordered credits with RA artist ids. */
    val lineup: List<LineupCredit>? = null,
    val interested: Int? = null,
    val attending: Int? = null,
    @SerialName("ra_url") val raUrl: String? = null,
    val image: String? = null,
    val description: String? = null,
    @SerialName("minimum_age") val minimumAge: Int? = null,
    @SerialName("venue_capacity") val venueCapacity: String? = null,
    val cost: String? = null,
) {
    val id: String get() = raEventId

    /** Credits in billing order, falling back to bare names on legacy rows. */
    val credits: List<LineupCredit>
        get() = lineup ?: artists.orEmpty().map { LineupCredit(null, it) }

    val visibleCredits: List<LineupCredit> get() = credits.take(4)
    val extraCredits: Int get() = (credits.size - 4).coerceAtLeast(0)

    /**
     * A door price only when the source actually gave one.
     *
     * Most rows are "0", "€" or blank, and nothing distinguishes "free" from
     * "unknown" — so anything without a non-zero digit is treated as unknown and
     * the chip is simply not shown. Guessing here would misstate what someone
     * pays at the door.
     */
    val entryLabel: String?
        get() {
            val raw = cost?.trim().orEmpty()
            if (raw.isEmpty()) return null
            return if (raw.any { it.isDigit() && it != '0' }) raw else null
        }

    /** "23:00" — the start, in venue time. */
    val startLabel: String?
        get() = runCatching {
            Instant.parse(startTime)
                .atZone(ZoneId.of("Europe/Madrid"))
                .format(DateTimeFormatter.ofPattern("HH:mm"))
        }.getOrNull()

    /**
     * Capacity, only when the column holds a real number.
     *
     * `venue_capacity` is free text on the source and carries everything from
     * "" to "500-1000" to prose. Anything that is not a plain positive integer
     * is dropped rather than printed — a capacity chip reading "TBC" is worse
     * than no chip.
     */
    val capacityLabel: String?
        get() = venueCapacity?.trim()?.toIntOrNull()?.takeIf { it > 0 }
            ?.let { "%,d".format(Locale.getDefault(), it) }

    /**
     * "23:00 – 06:00", and whether the end lands on the NEXT day.
     *
     * The median night here runs six hours and most end between 02:00 and
     * 08:00, so the end time usually belongs to tomorrow. Rendering it bare
     * reads as wrong, which is why the caller gets the flag as well as the text.
     */
    val timeRange: Pair<String, Boolean>?
        get() {
            val zone = ZoneId.of("Europe/Madrid")
            val start = runCatching { Instant.parse(startTime).atZone(zone) }.getOrNull()
                ?: return null
            val pattern = DateTimeFormatter.ofPattern("HH:mm")
            val end = runCatching { Instant.parse(endTime).atZone(zone) }.getOrNull()
                ?: return start.format(pattern) to false
            val crosses = start.toLocalDate() != end.toLocalDate()
            return "${start.format(pattern)} – ${end.format(pattern)}" to crosses
        }

    /** Day / month / weekday for the date block on the card. */
    val dateParts: Triple<String, String, String>
        get() = runCatching {
            val day = java.time.LocalDate.parse(date)
            Triple(
                day.format(DateTimeFormatter.ofPattern("EEE", Locale.getDefault())).uppercase(),
                day.dayOfMonth.toString(),
                day.format(DateTimeFormatter.ofPattern("MMM", Locale.getDefault())).uppercase(),
            )
        }.getOrElse { Triple("", "", "") }
}

/**
 * A DJ surfaced on a club page in place of a real event — the "this slot is a
 * DJ set, not an event" case. Flattens a `club_dj_sets` row joined to its `djs`
 * catalogue row.
 */
data class FeaturedDJ(
    val raArtistId: String,
    val name: String,
    val genres: List<String> = emptyList(),
    val instagram: String? = null,
    val soundcloud: String? = null,
    val website: String? = null,
    val knownVenues: List<String> = emptyList(),
    val regions: List<String> = emptyList(),
    val bio: String? = null,
    val raUrl: String? = null,
    val imageUrl: String? = null,
    val coverImageUrl: String? = null,
    val raFollowers: Int? = null,
    // From the club_dj_sets slot (curated, optional):
    val residencyLabel: String? = null,
    val night: String? = null,
) {
    val id: String get() = raArtistId

    /**
     * A DJ not in our RA catalogue — surfaced from a single-DJ night by name
     * only (synthetic "guest:" id). Rendered as a "Special guest".
     */
    val isGuest: Boolean get() = raArtistId.startsWith("guest:")

    /** Origin from the most-played region, when present. */
    val origin: String? get() = regions.firstOrNull()?.takeIf { it.isNotEmpty() }

    /** "Resident · Saturdays" — only the parts we actually have. */
    val residencyLine: String?
        get() = listOfNotNull(residencyLabel?.trim(), night?.trim())
            .filter { it.isNotEmpty() }
            .takeIf { it.isNotEmpty() }
            ?.joinToString(" · ")
}

/** A `djs` catalogue row. */
@Serializable
data class DjCatalogueRow(
    @SerialName("ra_artist_id") val raArtistId: String,
    val name: String,
    val genres: List<String>? = null,
    val instagram: String? = null,
    val soundcloud: String? = null,
    val website: String? = null,
    @SerialName("known_venues") val knownVenues: List<String>? = null,
    val regions: List<String>? = null,
    val bio: String? = null,
    @SerialName("ra_url") val raUrl: String? = null,
    @SerialName("image_url") val imageUrl: String? = null,
    @SerialName("cover_image_url") val coverImageUrl: String? = null,
    @SerialName("ra_followers") val raFollowers: Int? = null,
) {
    fun toFeaturedDj(residencyLabel: String? = null, night: String? = null) = FeaturedDJ(
        raArtistId = raArtistId,
        name = name,
        genres = genres.orEmpty(),
        instagram = instagram,
        soundcloud = soundcloud,
        website = website,
        knownVenues = knownVenues.orEmpty(),
        regions = regions.orEmpty(),
        bio = bio,
        raUrl = raUrl,
        imageUrl = imageUrl,
        coverImageUrl = coverImageUrl,
        raFollowers = raFollowers,
        residencyLabel = residencyLabel,
        night = night,
    )
}

/** The nested-join row: slot fields at top level, the DJ embedded. */
@Serializable
data class ClubDjSetRow(
    @SerialName("residency_label") val residencyLabel: String? = null,
    val night: String? = null,
    val dj: DjCatalogueRow? = null,
) {
    fun toFeaturedDj(): FeaturedDJ? = dj?.toFeaturedDj(residencyLabel, night)
}

/**
 * One dated appearance from `dj_appearances`.
 *
 * `club_id` is set only for venues we carry; rows in cities we have not
 * launched still come back (a DJ touring is signal, not a gap) and the UI
 * offers a "coming soon" note for that city rather than a dead link.
 */
@Serializable
data class DjGig(
    @SerialName("ra_event_id") val raEventId: String,
    val title: String? = null,
    val date: String = "",
    @SerialName("start_time") val startTime: String? = null,
    @SerialName("venue_name") val venueName: String? = null,
    val city: String? = null,
    val country: String? = null,
    @SerialName("club_id") val clubId: String? = null,
)
